File: 11.30/baekjoon_16566_박동현.java
/*
 * 같은 로직의 자바코드는 무리 없이 통과함
 */

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int M;
    static int[] cards;
    static boolean[] visit;
    static StringBuffer sb = new StringBuffer();
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());

        cards = new int[M];
        st = new StringTokenizer(br.readLine());
        for (int i=0; i<M; i++) {
            cards[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(cards);

        visit = new boolean[M];
        st = new StringTokenizer(br.readLine());
        for (int i=0; i<K; i++) {
            sb.append(bisect(Integer.parseInt(st.nextToken()))).append("\n");
        }
        System.out.println(sb);
    }

    private static int bisect(int card) {
        int left = 0;
        int right = M-1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (card < cards[mid]) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        while (visit[left]) {
                left ++;
        }
        visit[left] = true;
        return cards[left];
    }
}

File: 24/11/11.30/baekjoon_16566_박동현.py
'''
카드 게임

이분 탐색

이분 탐색을 하면서, 방문한 위치에 대해서는 값을 더 올려 방문하는 방식
근데 왜 python으로는 시간초과 나고 pypy로만 가능하지?
'''

N,M,K = map(int,input().split()) 
# 뽑은 카드, M: 카드 개수 ( 1 이상 N 이하 )
cards = sorted([*map(int,input().split())])

visit = [False] * (M)
# 철수가 i번째로 내는 카드의 번호 
for i in map(int,input().split()):
    left, right = 0, M-1

    while left <= right:
        mid = (left+right) // 2
        if i < cards[mid]:
            right = mid - 1
        else:
            left = mid + 1

    while visit[left]:
        left += 1
    visit[left] = True
    
    print(cards[left])